win_pkg: add msiexec override to enable selection of 32 or 64 msiexec.exe

alexproca opened the pull request:
### What does this PR do?

Adds possibility to specify 32 or 64 `msiexec.exe`

### What issues does this PR fix or reference?

I don't know if this is a known issue or not but it relates to #19137. The problem is more or less [described here](https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/)

### Tests written?

No. I tested it on my setup. I have an x64 only msi and the official version failed to install it without throwing any error. With this change everything is fine with my x64 only msi.

alexproca committed:
Add option to select 32 or 64 version of msiexec

Pass path to msiexec overloading msiexec parameter

salt-jenkins commented:
@alexproca, thanks for your PR! By analyzing the history of the files in this pull request, we identified @twangboy, @terminalmage and @jacobhammons to be potential reviewers.

damon-atkins commented:
Trying to understand why this is an issue?
You have pointed out
https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/
However this is a system issue not a salt issue?

#19137 is not related.  #19137 is about the salt-minion installer being a msi install instead of non-Microsoft exe.

Have you try fixing your issue following the instructions in https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/   rather than changing salt?

If this change went ahead a lot more parts of the code would need to be changed, I suspect.

A 64bit msi install can install a 32 bit program or a 64bit program.
A 32bit msi can install a 64bit program as well, I believe.

alexproca commented:
Hi Damon!


Thanks for interest in this issue.


1. Just this piece of code will change. Only the way of calling msiexec changes so it will not have side effects.


2. This is a salt problem because it relies on a system setting to select which msiexec (32 or 64) gets called. This change enables people to specify if they want x64 version of msiexec


2. 32 bit msiexec will not install 64 bit msi. will fail silently and the salt job trying to install will stall and will have to be killed.




Get Outlook for Android






From: Damon Atkins

Sent: Monday, April 17, 18:55

Subject: Re: [saltstack/salt] Select 32 or 64 msiexec.exe (#40716)

To: saltstack/salt

Cc: Alex Proca, Mention



Trying to understand why this is an issue?

You have pointed out

https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/

However this is a system issue not a salt issue?


#19137 is not related. #19137 is about the salt-minion installer being a msi install instead of non-Microsoft exe.


Have you try fixing your issue following the instructions in https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/ rather than changing salt?


If this change went ahead a lot more parts of the code would need to be changed, I suspect.


A 64bit msi install can install a 32 bit program or a 64bit program.

A 32bit msi can install a 64bit program as well, I believe.


—

You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub, or mute the thread.

damon-atkins commented:
If this is required install, remove and upgrade code would need to be updated.  I would suggest a function to return the correct msiexe string/path.

alexproca commented:
I implemented it inline for install only because it seems like 32 version of msiexec can uninstall a 64 bit msi but cannot install a 64 bit one (cannot read the package).

You are right though and is better to apply it consistently across module. 

Extracted it to a separate method and updated remove function too. Upgrade is not yet implemented for windows.

damon-atkins commented:
> because it seems like 32 version of msiexec can uninstall a 64 bit msi 
https://www.howtogeek.com/184374/how-to-fix-the-windows-installer-service-could-not-be-accessed-error-in-windows-7/

Have you tried this fix and then see if the 32 bit can install a 64 bit version?
It could be that the 32bit calls the 64bit msiexe if required.

alexproca commented:
Hi @damon-atkins !

I started this PR like this:

1. Tried to install a msi with salt (it was 64 bit only but I didn't know that) ✖

2. Tried to install it manually from GUI and it worked ✔

3. Tried to install it from command line with msiexec and failed with following error✖

<img src="https://www.howtogeek.com/wp-content/uploads/2014/03/00_lead_image_windows_installer_error.png.pagespeed.ce.HrK-GtIHbu.png"/>

4. Tried to install it with `C:\Windows\SystemWOW64\msiexec.exe` and it worked ✔

5. Tried to uninstall it with `C:\Windows\SystemWOW64\msiexec.exe` and it worked ✔

6. Harcoded path `C:\Windows\SystemWOW64\msiexec.exe` in `win_pkg.py` file on my environment and it worked ✔

7. Modified `win_pkg.py` to select msiexec with respect to `arch` setting (added possibility for the user to select required msiexec)

8. Created PR

damon-atkins commented:
And that is the same for another windows system?

In terms of arch
- x86_64
- x86_32

However the salt grain cpuarch is AMD64 and armv7lq
Need to keep in mind that Windows has now been ported to ARM cpu
I would suggest which is sort of along the lines of rpm and some installer names:
- x86_64
- x86_32
- arm_64

alexproca commented:
Is the same for each 64-bit Windows.

In terms of arch. This term can be confusing. 
Maybe it would be better to overload msiexec parameter like this:

```
{% set SYSTEM_ROOT = "%SystemRoot%" %}
myoldruntime:
  '1.0':
    full_name: 'My old runtime'
    installer: 'salt://win/binaries/my old runtime.msi'
    reboot: False
    msiexec: '{{SYSTEM_ROOT}}\SystemWOW64\msiexec.exe'
    install_flags: '/q'
    uninstall_flags: '/q REMOVE=ALL'

```

This way the default value will be False, if it is True we will rely on system and if is full path we will use that.

damon-atkins commented:
Do a search on win_pkg and ping/ask a few other people for input. Like @UtahDave @TheBigBear 
Cheers
Damon.

alexproca commented:
Thanks for your input @damon-atkins

I'il ping all 51 contributors to `win_pkg` to ask their opinion on msiexec parameter override explained above. 

@UtahDave
@terminalmage
@twangboy
@rallytime
@basepi
@s0undt3ch
@cvrebert
@hrumph
@thatch45
@cachedout
@attiasr
@jfindlay
@thematthopkins
@zigarn
@olliewalsh
@eliasp
@alexproca
@TheBigBear
@dnd
@mpreziuso
@morganwillcock
@jacobhammons
@SEJeff
@RossHammer
@twalk482
@whiteinge
@ryepup
@quantonganh
@mprymek
@otrempe
@m03
@lorengordon
@lcrees
@joehealy
@jmacfar
@jacksontj
@holmboe
@DmitryKuzmenko
@damon-atkins
@baijum
@alxbse
@abednarik

lorengordon commented:
Having never run into the issue, I'm not sure I really understand the problem.

But, taking it on faith that anything like this is really necessary, some questions/suggestions:

* Is there a way to test the msi to determine its arch and then "do the right thing"?
* How does this handle the case where the 64bit msiexec does not exist (because it's a 32-bit only system)?
* As a matter of style, I'd suggest removing `_string` from var and method names; it's not really adding any value in this case.

TheBigBear commented:
Having never run into the issue myself either, I'm not sure I really understand that there is a real problem? 
Are you sure that what you are running into is not a faulty msi, or some other factor?
Can you explain and demonstrate your problem with a publicly available msi and give step by step of what works and what doesn't and how and where salt comes into it?

alexproca commented:
@lorengordon 

1. I'm investigating if there can be determinated which version of msiexec should be run. I'm not sure if this can be done.
2. If the path will be passed as a string file existence `C:\Windows\SystemWOW64\msiexec.exe` will be checked first and if this does not exist will fallback to system's `msiexec.exe`
3. Agree with removing of `_string` bit

@TheBigBear 

I cannot make public the current msi but I will make some gif's with the problem and post them here.

morganwillcock commented:
I think this is probably what happens with a chained MSI package that is built for a different architecture; launching from a 64 bit process is writing a package to disk (bypassing the wow64 file system redirection) that the 32 bit second stage can't find. For 64 bit systems I think it's normal to build a 64 bit MSI package, or have a legacy .exe installer that actually includes 32 bit and 64 bit install packages and picks the right one.

It would probably be easiest to just use jinja templating to set the path to the msiexec that you want (you can check your arch grain, set a variable to the string based on that, and then render that as the parameter, and this doesn't affect any other app). Anything automated would probably be quite tricky, you'd have to look for custom actions and any environment variables they might be using, and if you start overriding the environment it's likely to be working against what the package author intended.

alexproca commented:
Thanks for input @lorengordon and @morganwillcock 

Updated code and overloaded msiexec setting. Now if one wants a custom msiexec can specify it like this.

```
{% set SYSTEM_ROOT = "%SystemRoot%" %}
myoldruntime:
  '1.0':
    full_name: 'My old runtime'
    installer: 'salt://win/binaries/my old runtime.msi'
    reboot: False
    msiexec: '{{SYSTEM_ROOT}}\SysWOW64\msiexec.exe'
    install_flags: '/q'
    uninstall_flags: '/q REMOVE=ALL'
```

damon-atkins commented:
Please change the PR title to begin with win_pkg: add msiexec override  Makes it easy to find it.

damon-atkins commented:
Go Go Jenkins!

damon-atkins commented:
@alexproca you need to fix the Jenkins issues before Salt Team will merge the code.

alexproca commented:
I backported all changes to 2016.11 and PR #40785 is merged now. 

@damon-atkins mentioned 
> changes in old versions are rolled forward until they get into develop

Summing this two up => this PR is no longer needed. Changes will propagate from 2016.11 into develop eventually.

Thanks everyone for support.
